### PR TITLE
[GO] Device Creation Helper

### LINKED ISSUE #1304: Catena device creation helpers
## 📖 Description
Create helpers to generate device definitions in a more intuitive way. We want to do this with function chaining that builds a proto as you chain function calls.
---

## 🎯 Goal / Outcome
You should be able to create a device with a param using the following syntax:

```catenaDevice := NewDevice(slot, name, vendor, version, serialNumber)```

Then add a param either with:

```catenaDevice = catenaDevice.withInt32Param(fqoid, name, constraint, value)```

OR

```catenaDevice.add(Int32Param(fqoid, name, constraint, value)```

Mandatory params are created in the device constructor, then additional fields are populated with "with*" or "add*" commands.
DeviceRequest endpoints just return either toProtoDevice(grpc) or toJSON(rest)

---

## ✅ Definition of Done (DoD)
- [ ] Feature is implemented and works as expected
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### LINKED ISSUE #1323: [Go] Consistent Proto Escape Hatch for Wrapper Structs
## 📖 Description

Wrapper structs in the Go SDK (e.g. `CatenaAsset`) encapsulate proto types behind unexported fields. Since these are short-lived value objects used by implementors to return data from registered handlers, the unexported field adds friction with no meaningful benefit — sophisticated callers can't directly construct or assign the underlying proto.

---

## 🎯 Goal / Outcome

All wrapper structs holding proto fields should export those fields directly, making them first-class and assignable — consistent with how Go's stdlib treats similar transfer objects.

---

## ✅ Definition of Done

- [ ] Audit all wrapper structs in catena for unexported proto fields
- [ ] Export the proto field on each (e.g. `Asset *protos.ExternalObjectPayload`)
- [ ] Remove or deprecate redundant getters like `GetProtoAsset()` where they only exist to work around the unexported field
- [ ] Update any internal code that accessed the unexported field directly (e.g. `TranscodeAssetPayload`)
- [ ] Unit tests cover constructing a wrapper via the exported field and round-tripping it
- [ ] No breaking changes to existing callers of the high-level API

---

## 🛠️ Implementation Notes

- Exporting the field is idiomatic Go for short-lived transfer/value objects — see `http.Request`, `http.Response`
- `CatenaAsset.GetProtoAsset()` only exists because `asset` is unexported; once the field is exported it becomes redundant
- `TranscodeAssetPayload` directly mutates `asset.asset` and must be updated when the field is renamed/exported

---

## 🔗 Related

- asset.go
- Redundant workaround: `GetProtoAsset()` in asset.go

---

## 🚧 Open Questions

- **Naming**: `Asset`, `Proto`, or `Payload`? Pick one convention and apply it across all wrapper types before implementing.

### LINKED ISSUE #1436: ProductStruct Helpers
## 📖 Description

The sdk can manage the mandatory Product Struct. Add a `Server.RegisterProductStruct(slot uint16, struct ProductStruct)` it can store the product for each slot.

GetDevice: pass to BL then overwrite the `product/*` fields in whatever BL returned if registered for that slot.
GetValue: before passing to BL, if its `product/*` and there is a product registered for the slot just return product, else call the handler
SetValue: same as GetValue, but return forbidden for the illegel write if registered
GetParam: same as GetValue
ParamInfo: same as GetValue

This removes the need for `NewDevice()` to deal with the product struct at all.

---

## 🎯 Goal / Outcome

---

## ✅ Definition of Done (DoD)

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### PR DESCRIPTION
**Pull request by nelgd**
## 📖 Description
Refactors the Go SDK device model around a fluent `catena.Device` builder and moves mandatory **product** struct handling into the SDK. Applications register product identity once per slot via `Server.RegisterProductStruct`; the server then injects the product param on `GetDevice`, serves `product/*` on `GetValue` and `ParamInfo`, and rejects writes to product fields on `SetValue`.

This removes the map-based device construction path (`ToDevice`, `Param.ToMap`, `Constraint.ToMap`) and updates examples (`hello_world`, `oneOfEverything`) and transports to use the new typed API. `GetParam` continues to stay in sync with `GetDevice` by rebuilding the device definition and resolving params via `lookupParam(device.Proto, fqoid)`.

---

## 🎯 Goal / Outcome
- Adopters build devices with `catena.NewDevice(slot)` and `With*` builders instead of `map[string]any` + `ToDevice`.
- The mandatory SMPTE product struct is SDK-managed and spec-compliant (`catena_sdk`, `catena_sdk_version` set by the SDK).
- Business logic no longer duplicates product handling across GetDevice, GetValue, ParamInfo, and SetValue handlers.
- `Asset` and `CommandResult` expose their underlying protos via exported `Proto` fields for consistency with `Device` and `Param`.

---

## ✅ Definition of Done (DoD)
- [x] `RegisterProductStruct` injects product on GetDevice and serves product/* on GetValue / ParamInfo
- [x] SetValue to product/* returns `StatusCodePermissionDenied`
- [x] `NewDevice(slot)` simplified; product identity removed from device constructor
- [x] Map-based conversion APIs removed (`ToDevice`, `ToMap`)
- [x] `hello_world` and `oneOfEverything` examples migrated to fluent Device API
- [x] GetParam handler updated to use `device.Proto` with `lookupParam`
- [x] `Asset` / `CommandResult` refactored to exported `Proto` fields
- [x] Unit tests added/updated (`product_test.go`, `product_server_test.go`, `roundtrip_test.go`, transport tests)

---

## 🛠️ Implementation Notes (Optional)
- New types/helpers in `product.go`: `ProductStruct`, `ProductParam`, `ProductValues`, `productValueForOid`, `productParamInfosForOid`.
- `Device` now exposes `Proto *protos.Device` directly (replaces private `device` field and `ToProtoDevice()`).
- `ParamInfosForRequest` now accepts `*Device` instead of `map[string]any`.
- `oneOfEverything` adds `registerProductStructs()` and removes product branches from value/param-info handlers; slot 0 param makers (`makeCounterParam`, `makeRunningParam`) remain the shared source of truth for GetDevice and GetParam.
- Transports (`grpc_transport.go`, `rest_transport.go`) updated to read `device.Proto` when marshaling responses.

---

## 🔗 Related Issues / Links

Closes issues
- #1436 
- #1323
- #1304 

---

### CODE REVIEW

File: sdks/go/pkg/catena/device.go [hunk 1]
```diff
@@ -63,52 +61,154 @@ const (
...
+// product struct (the "catena_sdk_version" field).
+const SDKVersion = "1.0.0"
+
+// CatenaSDKURL identifies the Catena SDK in the device's mandatory product
+// struct (the "catena_sdk" field).
+const CatenaSDKURL = "https://github.com/rossvideo/Catena"
+
+// Device wraps protos.Device and exposes a fluent builder API.
 type Device struct {
 	device *protos.Device
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go]: while we're churning in here, may as well rename to Proto: https://github.com/rossvideo/Catena/issues/1323

File: sdks/go/pkg/catena/device.go [hunk 2]
```diff
@@ -63,52 +61,154 @@ const (
 	DetailLevelUnset         DetailLevel = protos.Device_UNSET
 )
 
-// Device wraps protos.Device for device model handling
+// SDKVersion is the Catena Go SDK version reported in the device's mandatory
+// product struct (the "catena_sdk_version" field).
+const SDKVersion = "1.0.0"
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go]: This should come from VERSION.txt in the root. I have thoughts.... discuss next meeting

File: sdks/go/pkg/catena/device.go [hunk 3]
```diff
@@ -63,52 +61,154 @@ const (
...
+}
 
-	device := &protos.Device{}
-	if err := protojson.Unmarshal(jsonData, device); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal JSON to Device proto: %w", err)
+// WithCommand inserts command into the device's commands map, keyed by oid. The
+// command's proto is deep-copied. A nil command is ignored.
+func (cd *Device) WithCommand(oid string, command *Param) *Device {
+	if command == nil || command.Proto == nil {
+		logger.Warning("Device.WithCommand called with nil command; ignoring", "oid", oid)
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go:99]: hit these warning branches

File: sdks/go/pkg/catena/constraint_test.go [hunk 4]
```diff
@@ -257,29 +257,13 @@ func TestAlarmSeverityConstants(t *testing.T) {
 	}
 }
 
-func TestConstraintToMap_ForDeviceDefinition(t *testing.T) {
+func TestDevice_WithSharedConstraint(t *testing.T) {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/constraint_test.go]: should move to device_test.go

File: sdks/go/pkg/catena/param.go [hunk 5]
```diff
@@ -116,20 +79,14 @@ func NewParamString(value string) *Param {
 	}
 }
 
-func NewParamStruct(value map[string]any) *Param {
-	cp := &Param{
+// NewParamStruct creates a STRUCT param. Its fields are defined by attaching
+// sub-params with WithParam; each sub-param carries its own value.
+func NewParamStruct() *Param {
+	return &Param{
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param.go]: whoops

File: sdks/go/pkg/catena/device.go (hunk 2 above)
Comment (abrown-RV) [reply sdks/go/pkg/catena/device.go]:
```go
func sdkVersion() string {
    bi, ok := debug.ReadBuildInfo()
    if !ok {
        return "(unknown)"
    }
    for _, d := range bi.Deps {
        if d.Path == "github.com/rossvideo/catena/sdks/go" {
            return d.Version // resolved from the git tag / pseudo-version
        }
    }
    return "(devel)"
}

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 6]
```diff
@@ -251,124 +171,77 @@ func buildDeviceDefinition(slot uint16, counter *CounterState, state *ExampleSta
...
-		}, true
+
+		structNumber, _ := structExample["number"].(int32)
+		structText, _ := structExample["text"].(string)
+
+		device := catena.NewDevice(2,
+			productString(product, "name"),
+			productString(product, "vendor"),
+			productString(product, "version"),
+			productString(product, "serial_number")).
```
[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/examples/oneOfEverything/device_handler.go]: If you wanna introduce the protect struct helpers here to make warning like this go away, go for it.

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/examples/oneOfEverything/device_handler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nelgd) [reply sdks/go/examples/oneOfEverything/device_handler.go]: Like make a product struct?

[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/examples/oneOfEverything/device_handler.go]: https://github.com/rossvideo/Catena/issues/1436

File: sdks/go/pkg/catena/device.go [hunk 7]
```diff
@@ -63,52 +63,170 @@ const (
 	DetailLevelUnset         DetailLevel = protos.Device_UNSET
 )
 
-// Device wraps protos.Device for device model handling
+// sdkModulePath is the SDK's Go module path, used to locate the SDK's resolved
+// version in the running binary's build info.
+const sdkModulePath = "github.com/rossvideo/catena/sdks/go"
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go]: these should go in product_struct.go

File: sdks/go/pkg/catena/param.go (hunk 5 above)
Comment (abrown-RV) [reply sdks/go/pkg/catena/param.go]: Struct Params take the map still, that's not part of the ToMap cleanup

File: sdks/go/pkg/catena/product.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	ProductOidSerialNumber     = "serial_number"
+	ProductOidCatenaSDKVersion = "catena_sdk_version"
+	ProductOidCatenaSDK        = "catena_sdk"
+)
+
+// ProductParam builds the mandatory read-only "product" STRUCT param from p,
+// including the SDK-managed catena_sdk and catena_sdk_version sub-params. The
+// SDK uses this to seed the product param into the device on GetDevice.
+func ProductParam(p ProductStruct) *Param {
+	return NewParamStruct().
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go]: This will have to change when you revert the `NewParamStruct` mixup, but it sparked a very good design session resulting in: https://github.com/rossvideo/Catena/issues/1441 so thanks for that :D

File: sdks/go/pkg/catena/device.go [hunk 9]
```diff
@@ -63,52 +63,170 @@ const (
...
+	cd.Proto.DefaultScope = scope
+	return cd
 }
 
-// GetProtoDevice returns the underlying protos.Device
-func (cd Device) GetProtoDevice() *protos.Device {
-	return cd.device
+// ToJSON serializes the device to its REST JSON representation using proto
+// field names.
+func (cd Device) ToJSON() ([]byte, error) {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go]: belongs in json_utils.go

Comment (abrown-RV) [inline sdks/go/pkg/catena/device.go]: belongs in json_utils.go

File: sdks/go/pkg/catena/product.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+		ProductOidSerialNumber:     p.SerialNumber,
+		ProductOidCatenaSDKVersion: SDKVersion,
+		ProductOidCatenaSDK:        CatenaSDKURL,
+	}
+}
+
+// isProductOid reports whether fqoid targets the product struct or one of its
+// sub-fields (e.g. "product" or "product/name").
+func isProductOid(fqoid string) bool {
+	fqoid = strings.TrimPrefix(fqoid, "/")
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go]:
```suggestion
```

File: sdks/go/pkg/catena/product.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+type ProductStruct struct {
+	Name         string
+	Vendor       string
+	Version      string
+	SerialNumber string
+}
+
+// Sub-OIDs of the mandatory product struct param.
+const (
+	ProductOidName             = "name"
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go:100]: may as well include ProductOid as an export const here.

File: sdks/go/pkg/catena/product.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+// sub-fields (e.g. "product" or "product/name").
+func isProductOid(fqoid string) bool {
+	fqoid = strings.TrimPrefix(fqoid, "/")
+	return fqoid == productOid || strings.HasPrefix(fqoid, productOid+"/")
+}
+
+// productValueForOid resolves a product FQOID (the whole struct for "product",
+// or a single field for "product/<field>") to a Value from p.
+func productValueForOid(p ProductStruct, fqoid string) (Value, StatusResult) {
+	fqoid = strings.TrimPrefix(fqoid, "/")
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go]: no trims

File: sdks/go/pkg/catena/product.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+	return NewParamStruct().
+		WithReadOnly(true).
+		WithAccessScope(ScopeMon).
+		WithParam(ProductOidName, NewParamString(p.Name)).
+		WithParam(ProductOidVendor, NewParamString(p.Vendor)).
+		WithParam(ProductOidVersion, NewParamString(p.Version)).
+		WithParam(ProductOidSerialNumber, NewParamString(p.SerialNumber)).
+		WithParam(ProductOidCatenaSDKVersion, NewParamString(SDKVersion)).
+		WithParam(ProductOidCatenaSDK, NewParamString(CatenaSDKURL))
+}
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go:128]: somewhere there should be a test that this creates the correct proto full up. and once you've got it maybe that's in all of the product_server_test.go tests, instead of poking at the return and verifying a few fields. The following I believe is the full correct proto. Might not be quite right but I think it is:
```go
	if !proto.Equal(product, &protos.Param{
		Type:        protos.ParamType_STRUCT,
		ReadOnly:    true,
		AccessScope: ScopeMon,
		Params: map[string]*protos.Param{
			ProductOidName:             {Type: protos.ParamType_STRING},
			ProductOidVendor:           {Type: protos.ParamType_STRING},
			ProductOidVersion:          {Type: protos.ParamType_STRING},
			ProductOidSerialNumber:     {Type: protos.ParamType_STRING},
			ProductOidCatenaSDKVersion: {Type: protos.ParamType_STRING},
			ProductOidCatenaSDK:        {Type: protos.ParamType_STRING},
		},
		Value: &protos.Value{
			Kind: &protos.Value_StructValue{
				StructValue: &protos.StructValue{
					Fields: map[string]*protos.Value{
						ProductOidName:             {Kind: &protos.Value_StringValue{StringValue: "Camera"}},
						ProductOidVendor:           {Kind: &protos.Value_StringValue{StringValue: "Ross Video"}},
						ProductOidVersion:          {Kind: &protos.Value_StringValue{StringValue: "1.0"}},
						ProductOidSerialNumber:     {Kind: &protos.Value_StringValue{StringValue: "SN-12345"}},
						ProductOidCatenaSDKVersion: {Kind: &protos.Value_StringValue{StringValue: SDKVersion}},
						ProductOidCatenaSDK:        {Kind: &protos.Value_StringValue{StringValue: CatenaSDKURL}},
					},
				},
			},
		},
	}) {
		t.Errorf("SDK-injected product param does not match expected, got %v", product)
	}
```
Maybe thats a good const in the testing code and then you can compare the full thing in getdevice/full getparam/full getvalue

File: sdks/go/pkg/catena/server.go [hunk 14]
```diff
@@ -198,6 +198,13 @@ type Server interface {
 	RegisterHeartbeatHandler(slot uint16, handler HeartbeatHandler)
 	RegisterAccessHandler(handler AccessHandler)
 
+	// RegisterProductStruct hands the mandatory product struct for a slot to the
+	// SDK. Once registered, the SDK injects the product param into the device on
+	// GetDevice, answers GetValue and ParamInfo for product/*, and rejects
+	// SetValue writes to product/* with StatusCodePermissionDenied — business
+	// logic no longer needs to handle the product struct.
+	RegisterProductStruct(slot uint16, product ProductStruct)
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go:213]: add a note for what happens if they don't register for a slot

File: sdks/go/pkg/catena/server.go [hunk 15]
```diff
@@ -614,6 +623,26 @@ func (s *server) RegisterAccessHandler(handler AccessHandler) {
...
+		s.notifySlotsAdded(slot)
+	}
+}
+
+// productForSlot returns the SDK-managed product struct registered for a slot,
+// if any.
+func (s *server) productForSlot(slot uint16) (ProductStruct, bool) {
+	s.mu.Lock()
+	product, ok := s.productStructs[slot]
+	s.mu.Unlock()
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]:
```suggestion
	defer s.mu.Unlock()
	product, ok := s.productStructs[slot]
```

File: sdks/go/pkg/catena/server.go [hunk 16]
```diff
@@ -632,7 +661,15 @@ func (s *server) InvokeGetDeviceHandler(slot uint16, transportContext TransportC
 	s.mu.Unlock()
 
 	if ok {
-		return handler(slot, handlerContext)
+		device, res := handler(slot, handlerContext)
+		// Overwrite the product/* fields in whatever the business logic returned
+		// with the SDK-managed product struct, if one is registered for the slot.
+		if res.Code == StatusCodeOk {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]:
```suggestion
		if res.IsOk() {
```

File: sdks/go/pkg/catena/server.go [hunk 17]
```diff
@@ -706,6 +751,16 @@ func (s *server) InvokeSetValueHandler(slot uint16, entries []SetValueEntry, tra
 		return StatusWithCode(StatusCodePermissionDenied, "Permission denied")
 	}
 
+	// The SDK-managed product struct is read-only; reject any write targeting
+	// product/* when a product is registered for the slot.
+	if _, has := s.productForSlot(slot); has {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]: Should reject regardless of if there's a struct registered or not. Even if BL is managing their own product, they still shouldn't accept writes, so just force it on them.

File: sdks/go/pkg/transports/README.md [hunk 18]
```diff
@@ -83,6 +92,8 @@ Both transports invoke the same registered handlers from `catena.ServerRuntime`:
 - `RegisterParamInfoHandler`
 - `RegisterHeartbeatHandler`
 
+The mandatory product struct is managed by the SDK, not by a handler: call `RegisterProductStruct(slot, catena.ProductStruct{...})` and the SDK injects it into GetDevice and serves product/* on GetValue and ParamInfo (rejecting writes).
```
Comment (abrown-RV) [inline sdks/go/pkg/transports/README.md]: include messaging for fallback, BL can still do their own if they really want to.

File: sdks/go/pkg/catena/param_info.go [hunk 19]
```diff
@@ -89,119 +89,51 @@ func ToParamInfo(m map[string]any) (ParamInfo, error) {
...
-	params, ok := deviceDefinition["params"].(map[string]any)
-	if !ok {
-		return []ParamInfo{}, StatusWithCode(StatusCodeInternal, "invalid device params structure")
+// a Device's params subtree.
+func ParamInfosForRequest(fqoid string, device *Device, recursive bool) ([]ParamInfo, StatusResult) {
+	if device == nil || device.Proto == nil {
+		return []ParamInfo{}, StatusWithCode(StatusCodeInternal, "invalid device")
 	}
 
 	oid := strings.TrimPrefix(fqoid, "/")
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_info.go]: clean up this old trim while you're in here.

File: sdks/go/pkg/catena/param_info.go
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_info.go]: probably don't want this anymore either

File: sdks/go/examples/oneOfEverything/device_handler.go [hunk 20]
```diff
@@ -5,115 +5,78 @@ import (
...
 // buildDeviceDefinition returns the descriptor for one slot. It is a function
 // (not a static YAML file) so every param's value field reflects live state at
-// GetDevice time. device_handler.go calls this and passes the result to
-// catena.ToDevice. Keep param OIDs in sync with value_handlers and param_info_handler.
+// GetDevice time. device_handler.go calls this and returns the result directly.
+// Keep param OIDs in sync with value_handlers and param_info_handler.
+//
+// catena.NewDevice seeds the mandatory "product" struct param from the
+// identity fields passed to it; additional params, commands, constraints, and
+// menus are attached with the fluent With* builders.
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/device_handler.go]: not anymore

**Review by abrown-RV — CHANGES_REQUESTED** [review]
good work Nelson

File: sdks/go/pkg/catena/param.go [hunk 21]
```diff
@@ -116,12 +79,18 @@ func NewParamString(value string) *Param {
...
+// NewParamStruct creates a STRUCT param whose value is built from the given
+// field map. Field descriptors may additionally be attached with WithParam.
+// A nil or empty map yields a descriptor-only struct with no value set.
 func NewParamStruct(value map[string]any) *Param {
 	cp := &Param{
 		Proto: &protos.Param{
 			Type: protos.ParamType_STRUCT,
 		},
 	}
+	if len(value) == 0 {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param.go]: revert this too, this shouldn't change at all in this PR

File: sdks/go/pkg/catena/param_test.go [hunk 22]
```diff
@@ -83,6 +83,20 @@ func TestNewParamStruct(t *testing.T) {
 	}
 }
 
+func TestNewParamStruct_WithValue(t *testing.T) {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_test.go]: this test already existed below, just revert that one

File: sdks/go/pkg/catena/param_test.go [hunk 23]
```diff
@@ -883,16 +889,6 @@ func TestWithParam_SelfReference(t *testing.T) {
 
 // --- Error path tests ---
 
-func TestNewParamStruct_InvalidValue(t *testing.T) {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_test.go:886]: restore this test too

File: sdks/go/pkg/catena/param_test.go [hunk 24]
```diff
@@ -699,21 +713,13 @@ func TestSetGetValue_Roundtrip_Struct(t *testing.T) {
...
-}
-
 func TestNewParamStruct_NoValue(t *testing.T) {
 	p := NewParamStruct(nil).Proto
 	if p.GetType() != protos.ParamType_STRUCT {
 		t.Errorf("expected STRUCT, got %v", p.GetType())
 	}
 	if p.GetValue() != nil {
-		t.Error("expected nil value when no arg provided")
+		t.Error("expected nil value for a descriptor-only struct (nil map)")
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/param_test.go]: revert

File: sdks/go/pkg/catena/product.go [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+		field := strings.TrimPrefix(fqoid, ProductOid+"/")
+		value, ok := values[field]
+		if !ok {
+			return ReplyError[Value](StatusCodeNotFound, "parameter not found: "+fqoid)
+		}
+		native = value
+	}
+
+	value, res := ToValue(native)
+	if res.Code != StatusCodeOk {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/product.go:168]: FYI there are shortcut `StatusResult.IsOk()` and `StatusResult.IsError()` funcs now. Not saying go back and change *everything*, but try and remember going forward

File: sdks/go/pkg/transports/json_utils.go [hunk 26]
```diff
@@ -118,6 +118,16 @@ func MarshalAssetJSON(asset *protos.ExternalObjectPayload) ([]byte, error) {
 	return protoMarshalOpts.Marshal(asset)
 }
 
+// DeviceToJSON serializes a device to its REST JSON representation using proto
+// field names, without the SMPTE-compliant cleanup that MarshalDeviceJSON
+// applies.
+func DeviceToJSON(device *catena.Device) ([]byte, error) {
```
Comment (abrown-RV) [inline sdks/go/pkg/transports/json_utils.go]: unnecessary, remove, sorry, wasn't clear until it was moved that actually we just don't want it anymore

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/catena/product.go (hunk 25 above)
Comment (nelgd) [reply sdks/go/pkg/catena/product.go:168]: Sounds good, will keep this in mind

**Comment by abrown-RV** [comment]
merge with develop

**Review by abrown-RV — APPROVED** [review]
